### PR TITLE
Refactor ReaDDy processing and all visualization

### LINKED ISSUE #21: Simultaneous visualization of sub-sampled cytosim and readdy simulations
### Feature Description
We want to simultaneously visualize sub-sampled readdy and cytosim simulations using appropriate colorization and metric plots

### Use Case
This will allow us to visually compare simulations across simulators and repeats.

### Solution
1. Each simularium trajectory should correspond to all repeats of cytosim and readdy for a particular velocity
2. This can be achieved by using `AgentData` append method in `simulariumio`. See `notebooks/create_simularium_outputs.py` for previous efforts in this direction.

### LINKED ISSUE #25: Import actin coordinates from tomography data and visualize and calculate metrics

### PR DESCRIPTION
**Pull request by blairlyons**
Time Estimate or Size
=======
x-large (sorry!)

Problem
=======
Visualization and post processing scripts were inefficiently organized and out of date. 

Solution
========
I refactored code and tested all of the notebooks in simulation.readdy and visualization modules.

## Type of change
Please delete options that are not relevant.

* Breaking change (fix or feature that would cause existing functionality to not work as expected)
* This change requires a documentation update

Keyfiles:
-----------------------
Main changes are in simulation.readdy and visualization modules, but I had to make some changes to analysis and other places to support these.

### CODE REVIEW

File: subcell_pipeline/analysis/compression_metrics/README.md
Comment (jessicasyu) [inline subcell_pipeline/analysis/compression_metrics/README.md]:
```suggestion
Analysis combines compression simulations from Cytosim and Readdy and calculates various compression metrics to compare fibers.
```

File: subcell_pipeline/analysis/dimensionality_reduction/fiber_data.py
Comment (mogres) [inline subcell_pipeline/analysis/dimensionality_reduction/fiber_data.py]: I think this is supposed to be the y axis?

Comment (jessicasyu) [inline subcell_pipeline/analysis/dimensionality_reduction/fiber_data.py]: Positive y axis is correct here. The function uses the yz plane and rotates that "loop" to align onto the y axis

File: subcell_pipeline/analysis/tomography_data/_analyze_actin_cme_tomography_data.py
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/_analyze_actin_cme_tomography_data.py]: I would leave this as a configuration parameter instead of a constant since it isn't imported anywhere else and depends on the data set.

File: subcell_pipeline/analysis/tomography_data/tomography_data.py
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: Can we have the return be a regular `bool`? I think you can just have the return be `np.all(dot_products > 0).item()` and it should work.

File: subcell_pipeline/analysis/tomography_data/tomography_data.py
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: nit: replace return type with `:` so you don't need to keep two type hints in sync (the autodoc will grab the one in the method signature)

File: subcell_pipeline/analysis/tomography_data/tomography_data.py
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: You can use `io_collection.save.save_figure` here to save the `fig` to a local path or S3 bucket. I think you want to move the figure saving into the loop (and update the output key) so you save each individual figure, rather than just the final figure.

File: subcell_pipeline/simulation/readdy/loader.py
Comment (jessicasyu) [inline subcell_pipeline/simulation/readdy/loader.py]: nit: Absolute instead of relative import

File: subcell_pipeline/visualization/_visualize_tomography_data.py
Comment (jessicasyu) [inline subcell_pipeline/visualization/_visualize_tomography_data.py]:
```suggestion
- [Visualize tomography data](#visualize-tomography-data)
```

File: subcell_pipeline/visualization/_visualize_cytosim_trajectories.py
Comment (jessicasyu) [inline subcell_pipeline/visualization/_visualize_cytosim_trajectories.py:87]: I was able to run this successfully (just for the first three seeds of the `0047` condition) but the linkers seem to be misaligned with the actual fiber.

**Review by jessicasyu — COMMENTED** [review]
Looks great! Mostly minor things. The only major issue is with the Cytosim simularium trajectories, where the linkers seem to be off.

There's also some missing docstrings, but we can add those in a separate PR so this one doesn't keep getting bigger ;)

File: subcell_pipeline/simulation/readdy/_process_readdy_compression_simulations.py
Comment (mogres) [inline subcell_pipeline/simulation/readdy/_process_readdy_compression_simulations.py]: should these be calculated based on the velocity and the compression distance for consistency instead of being hard coded?

File: subcell_pipeline/simulation/readdy/loader.py
Comment (mogres) [inline subcell_pipeline/simulation/readdy/loader.py:62]: Does this function need documentation or the class attribute descriptions are sufficient?

File: pyproject.toml
Comment (mogres) [inline pyproject.toml]: Would recommend adding packages using pdm by running `pdm add imageio`. This will also update the lockfile correctly for `pdm sync`

**Review by mogres — COMMENTED** [review]
Thanks for making the changes, the visualization updates are very helpful! The PR overall looks good to me. I was able to run the visualization notebooks and generate the simularium trajectories but ran into a couple of issues along the way.

1) In the individual readdy and cytosim visualization notebooks, I ran into an error due to the missing `h5py` package. After installing it using `pdm add h5py` the notebooks ran as expected.

2) I also ran into the issue Jess mentioned regarding the linkers in Cytosim trajectories. They seem to be offset from the main fiber.
image

File: subcell_pipeline/analysis/tomography_data/tomography_data.py
Comment (blairlyons) [reply subcell_pipeline/analysis/tomography_data/tomography_data.py]: oh I tried to use save image but had to avoid it because I didn't want to depend on bioio for the OmeTiffWriter. It would help if this package was documented :)

Comment (jessicasyu) [reply subcell_pipeline/analysis/tomography_data/tomography_data.py]: Haha yes I've been slowly working on the docs :)

The default install of `io-collection` does not include the `bioio` install, so should be okay.

File: subcell_pipeline/simulation/readdy/loader.py
Comment (blairlyons) [reply subcell_pipeline/simulation/readdy/loader.py:62]: since it's the constructor I usually just rely on the class documentation

File: subcell_pipeline/visualization/_visualize_cytosim_trajectories.py
Comment (blairlyons) [reply subcell_pipeline/visualization/_visualize_cytosim_trajectories.py:87]: I forgot, I had to make a small change to simulariumio to fix this. It's in [PR](https://github.com/simularium/simulariumio/pull/188) now and will hopefully be released shortly

**Comment by blairlyons** [comment]
@jessicasyu @mogres I think I addressed everything, the Cytosim issue depends on [this PR](https://github.com/simularium/simulariumio/pull/188) getting merged. I tried fixing the mypy error in the analysis module without success, would be great if someone knows how to fix that.

File: subcell_pipeline/analysis/tomography_data/tomography_data.py
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: I think you want to save separate figures for each dataset?

File: .github/workflows/build.yml
Comment (jessicasyu) [inline .github/workflows/build.yml:61]: I would keep the build workflow as is (it'll break the coverage badge otherwise). It just won't find any tests to run, which I think is fine.

File: .pdm-python
Comment (jessicasyu) [inline .pdm-python]: We shouldn't commit this file -- probably add to the `.gitignore`?

File: subcell_pipeline/simulation/readdy/parser.py
Comment (jessicasyu) [inline subcell_pipeline/simulation/readdy/parser.py]: Needs docstrings (or make private with `_` prefix and they won't show up in the docs)

**Review by jessicasyu — APPROVED** [review]
Looks great! Just a couple more things but good to merge after that. I can take a final pass through everything once we get the PRs all merged.

**Comment by jessicasyu** [comment]
@blairlyons @mogres I believe that last mypy issues is fixed in #61

**Review by mogres — APPROVED** [review]

File: .github/workflows/build.yml
Comment (blairlyons) [reply .github/workflows/build.yml:61]: it breaks the build otherwise. We can add one test that passes, or maybe the coverage badge should be removed since we know it's zero, unfortunately

File: .pdm-python
Comment (blairlyons) [reply .pdm-python]: whoops

File: .github/workflows/build.yml
Comment (jessicasyu) [reply .github/workflows/build.yml:61]: I can refactor the existing test so it passes? It doesn't need to be in this PR

Comment (blairlyons) [reply .github/workflows/build.yml:61]: can I leave it like this for this PR and you can pull this back in with the fixed test?

**Pull request merged**
